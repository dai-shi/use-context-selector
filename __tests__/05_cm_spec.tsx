import React, {
  Dispatch,
  ReactNode,
  SetStateAction,
  useRef,
  useState,
  StrictMode,
} from 'react';

import {
  render, fireEvent, cleanup, screen,
} from '@testing-library/react';

import {
  createContext,
  useContextSelector,
  useContextUpdate,
} from '../src/index';

describe('useContextUpdate spec', () => {
  afterEach(cleanup);

  it('counter', () => {
    const initialState = {
      count1: 0,
      count2: 0,
    };
    type State = typeof initialState;
    const context = createContext<[
      State,
      Dispatch<SetStateAction<State>>,
    ]>(
      [initialState, () => null],
    );
    const Counter1 = () => {
      const count1 = useContextSelector(context, (v) => v[0].count1);
      const setState = useContextSelector(context, (v) => v[1]);
      const update = useContextUpdate(context);
      const increment = () => update(() => {
        setState((s) => ({
          ...s,
          count1: s.count1 + 1,
        }));
      });
      const renderCount = useRef(0);
      renderCount.current += 1;
      return (
        <div>
          <span>count1: {count1}</span>
          <button type="button" onClick={increment}>+1</button>
          <span>{renderCount.current}</span>
        </div>
      );
    };
    const Counter2 = () => {
      const count2 = useContextSelector(context, (v) => v[0].count2);
      const renderCount = useRef(0);
      renderCount.current += 1;
      return (
        <div>
          <span>count2: {count2}</span>
          <span data-testid="counter2">{renderCount.current}</span>
        </div>
      );
    };
    const StateProvider = ({ children }: { children: ReactNode }) => (
      <context.Provider value={useState(initialState)}>
        {children}
      </context.Provider>
    );
    const App = () => (
      <StrictMode>
        <StateProvider>
          <Counter1 />
          <Counter2 />
        </StateProvider>
      </StrictMode>
    );
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByTestId('counter2').textContent).toEqual('1');
    expect(container).toMatchSnapshot();
  });
});
