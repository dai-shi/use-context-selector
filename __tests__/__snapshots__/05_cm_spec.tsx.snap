// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContextUpdate spec counter 1`] = `
<div>
  <div>
    <span>
      count1: 
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span>
      1
    </span>
  </div>
  <div>
    <span>
      count2: 
      0
    </span>
    <span
      data-testid="counter2"
    >
      1
    </span>
  </div>
</div>
`;

exports[`useContextUpdate spec counter 2`] = `
<div>
  <div>
    <span>
      count1: 
      1
    </span>
    <button
      type="button"
    >
      +1
    </button>
    <span>
      5
    </span>
  </div>
  <div>
    <span>
      count2: 
      0
    </span>
    <span
      data-testid="counter2"
    >
      1
    </span>
  </div>
</div>
`;
